kinvolk/inspektor-gadget
# dns gadget: rework tracer with networkHost pods

# dns gadget: rework tracer with networkHost pods

Don't check network namespace (/proc/pid/ns/net) of terminated containers. For this, gadgettracermanager now adds netns in containerDefinition.

The tracer keeps a reference count when it is attached on a network namespace. gadget.go might attempt to attach the tracer on the same netns several times because:

1. a pod might have several containers
2. a pod might run with networkHost=true

In those cases, the tracer is added only one time, and it is detached when the reference counter reaches zero.

Additionally, error and notice messages are improved.

## How to use

[ describe what reviewers need to do in order to validate this PR ]

## Testing done

- `./kubectl-gadget dns --verbose -A`
  - Check pods with networkHost=true
  - Check pods with several containers
- Stop the tracer and check for leaks:
  - Check the file descriptor list of thegadgettracermanager process (`/proc/$pid/fd`): the bpf program and maps for the gadget should have been removed
  - Check the go routine list of thegadgettracermanager process (`gadgettracermanager -dump`): go routines for gadgets should not be present anymore
